WaveformExtractor -> SortingAnalyzer doc page

chrishalcrow opened the pull request:
@samuelgarcia  and @alejoe91  suggested making a temporary page to help people convert their WaveformExtractor code to SortingAnalyzer code. Here’s my attempt at that page.

When you build the documentation, there should be a new “From WaveformExtractor to SortingAnalyzer” page in the "How to Guides".

It comes in three sections:
1. Why change?
2. Converting existing folders
3. A “code dictionary”

Happy to hear feedback on the overall structure, any useful functions I’ve missed out, or anything else helpful to include! Also, feel free to completely rewrite the "Why Change?" section - this was my best summary based on old discussions :)

Note: this requires `sphinx-design’ to build, for the two-column code. You just need to pip install sphinx-design. When installed it should look like this:

![Screenshot from 2024-03-26 17-01-39](https://github.com/SpikeInterface/spikeinterface/assets/57948917/7c128b1e-f904-4b7c-8d8d-a32112d0783c)

zm711 commented:
I will read over later, but I think this is a brilliant idea :)

alejoe91 commented:
For the failing docs build, you can just add `sphinx-design` here: https://github.com/chrishalcrow/spikeinterface/blob/sorting_analyzer_VS_waveform_extractor/pyproject.toml#L159

alejoe91 commented:
Couple of quick comments:

- can you change `extractor` to `waveform_extractor`?
- can you use full black format? E.g.:
```
waveform_extractor = extract_waveforms(
    recording,
    sorting,
    ....
)
```
- let's use keyword arguments for clarity
- `sorter` --> `sorting`

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+Why change?
+^^^^^^^^^^^
+
+Previously, a lot of post-processing was handled by the ``WaveformExtractor`` class. However, as the name suggests, this singles out waveforms as special. But many outputs don't care about waveforms. And this isn't just a naming problem: you had to create a WaveformExtractor object (and thus calculate waveforms) to calculate e.g. spike_amplitudes, which didn't necessarily depend on them. Then to  ...
```
spike_amplitudes does require knowledge of the waveforms :) but to get around this we have fast_templates which still requires using the random_spikes to get the info without officially calculating the waveforms. Maybe we pick something that is not an edge case and use correlograms as an example here since that really only requires the spike times?

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+* `Quality Metrics <#id5>`_
+
+Throughout this section, we assume that all functions have been imported into the namespace.
+E.g. we have run code such as ``from spikeinterface.postprocessing import extract_waveforms`` for each function. If you have imported
```
Where did you find this? I don't think this exists? 

```python
ImportError: cannot import name 'extract_waveforms' from 'spikeinterface.postprocessing
```

chrishalcrow replied:
Good spot - it exists in the pre-SortingAnalyzer spikeinterface versions. Best not to advertise them!

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+                recording = recording)
+
+
+By default, the object is stored in memory. Alternatively, we can save it locally at the point of creation by specifying a ``folder`` and a ``format``. Additionally, you can decide whether to use sparsity or not
```
I would add a note here that in memory means that if you end the session before running `save_as` everything will be lost. Nothing worse then forgetting to save and losing a bunch of work :) You have the section below, but the goal according to Sam is in-memory is for testing conditions and iterating then saving. So actually explaining the intention of memory vs not is another change between the waveform extractor and the sorting analyzer

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+            analyzer.get_dtype()
+            analyzer.is_sparse()
+
+
```
Are we missing on the section where the waveform_extractor required the recording or needed to be put into recordingless mode whereas the sorting_analyzer automatically deals with this?

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+            wv_data = wv.get_data()
+            ul = analyzer.get_extension(
+                extension_name = "unit_locations")
+            ul_data = nl.get_data()
```
```suggestion
            ul_data = ul.get_data()
```

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+
+
+
+You can also access the parameters used in the extension calculation, which is very simply for the new SortingAnalyzer:
```
```suggestion
You can also access the parameters used in the extension calculation, which is very simple for the new SortingAnalyzer:
```

zm711 reviewed (commented):
First set of my comments/questions.

chrishalcrow commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+
+
+You can also find some fundamental properties of the object,
+though these are mostly used internally:
```
It's tricky. My feeling is that this isn't a tutorial or learning guide; more of a reference guide. Hence, I'm tempted to throw everything in it.

chrishalcrow commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+
+
+
+You can delete extensions. Note that if you delete a child all of its parents
```
I've done everything in reverse. Oops.

zm711 replied:
Okay cool. That's what I thought.

zm711 commented:
extractor is overloaded in the code base which is why it would add confusion to just say extractor (in my opinion). In some of the examples we switched to `wvf_extractor` (or the much worse for clarity `we`). I do agree the scroll bar turns people off of that part of the docs. So opinions on `wvf_extractor`?

alejoe91 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,572 @@
+                recording = recording,
+                folder = "my_sorting_analyzer",
+                format = "binary_folder",
+                sparse = True)
```
I think that's too much ;)

alejoe91 commented:
Any idea why it's not shouwing up in the docs build here? https://spikeinterface--2633.org.readthedocs.build/en/2633/

After fixing this, I would add a reference to this page to the "warning" here: https://github.com/SpikeInterface/spikeinterface/blob/597adde034a29b329e1b4181ea81db77f7799424/doc/index.rst

chrishalcrow commented:
@alejoe91 @zm711 I had linked to a folder, rather than the rst file. Should work now. Also changed `extractor` to `wv_extractor` and fixed some typos.

I'm going hiking till Monday - I won't be offended if you make big changes without me: probably good to get this published soon!

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+            wvf_extractor = extract_waveforms(
+                sorting=sorting,
+                recording=recording,
+                mode="folder",
+                folder="my_waveform_extractor",
```
```suggestion
                folder="my_waveform_extractor",
                mode="folder",
```
Just so folder lines up between the two objects.

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+                recording=recording
+            )
+
+By default, the object is stored in memory. In this case, if you end your session without saving (which you can do using `save_as`, see below) you'll lose everything!
```
```suggestion
By default, the object is stored in memory. In this case, if you end your session without saving (which you can do using ``save_as``, see below) you'll lose everything!
```
Italicized or code-format?

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+++++++++++++++++++
+
+Waveforms, templates, quality metrics etc are all extensions of the ``SortingAnalyzer`` object.
+Some extensions depend on other extensions. To calculate a *child* we must first have calculated it's
```
```suggestion
Some extensions depend on other extensions. To calculate a *child* we must first have calculated its
```

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+            )
+
+Read more about extensions and their keyword arguments in the
+`PostProcessing module documentation <https://spikeinterface.readthedocs.io/en/latest/modules/postprocessing.html>`_
```
```suggestion
`postprocessing module documentation <https://spikeinterface.readthedocs.io/en/latest/modules/postprocessing.html>`_
```

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+
+        .. code-block:: python
+
+            # This also deletes any parents
```
```suggestion
            # This also deletes any children
```

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+            ul = analyzer.get_extension(
+                extension_name="unit_locations"
+            )
+            ul_data = nl.get_data()
```
```suggestion
            ul_data = ul.get_data()
```

This fix just got lost in one of the commits :)

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+            )
+            ul_data = nl.get_data()
+
+You can also access the parameters used in the extension calculation, which is very simply for the new SortingAnalyzer:
```
```suggestion
You can also access the parameters used in the extension calculation, which is very simple for the new SortingAnalyzer:
```
This one too. Probably just forgot to pull from the github changes so these were overwritten when you pushed.

zm711 commented on doc/how_to/waveform_extractor_to_sorting_analyzer.rst:
```diff
@@ -0,0 +1,566 @@
+
+        .. code-block:: python
+
+            ul_parms = ul.params
```
```suggestion
            ul_params = ul.params
```

zm711 commented on doc/index.rst:
```diff
@@ -17,8 +17,9 @@ state-of-the-art spike sorters, post-process and curate the output, compute qual
-    Please refer to the stable documentation `here <https://spikeinterface.readthedocs.io/en/0.100.1>`_
+    `enhancement proposal <https://github.com/SpikeInterface/spikeinterface/issues/2282>`_. Learn how to update your code
+    `here <how_to/waveform_extractor_to_sorting_analyzer.html>`_ and read more about the
+    :code:`SortingAnalyzer` `here <modules/postprocessing.html>`_.
```
The SortingAnalyzer itself lives in core. So not sure about this line. I think linking the `how_to` is great, though.

zm711 reviewed (commented):
Enjoy hiking! I did another careful proofread just for typos this time. Once these are done it looks good to me :) And I checked and it built this time perfectly.

samuelgarcia commented:
Merci Chris. this is impressive.

chrishalcrow commented:
Thanks for the careful readthrough @zm711! And thanks for merging it @samuelgarcia - hope it's helpful.
